pkg/kubectl/cmd/exec.go

```diff
@@ -174,16 +174,28 @@ func (p *ExecOptions) Run() error {
 		return err
 	}
 
-	if pod.Status.Phase != api.PodRunning {
-		return fmt.Errorf("pod %s is not running and cannot execute commands; current phase is %s", p.PodName, pod.Status.Phase)
-	}
-
 	containerName := p.ContainerName
 	if len(containerName) == 0 {
 		glog.V(4).Infof("defaulting container name to %s", pod.Spec.Containers[0].Name)
 		containerName = pod.Spec.Containers[0].Name
 	}
 
+	isInitContainer := false
+	for _, initContainer := range pod.Spec.InitContainers {
+		if initContainer.Name == containerName {
+			isInitContainer = true
+			break
+		}
+	}
+
+	if !isInitContainer && pod.Status.Phase != api.PodRunning {
```

Comment:
The Kubelet should be able to return structured errors - check out
kubelet.go#validateContainerLogStatus() for some examples.


On Thu, Jun 2, 2016 at 8:31 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/kubectl/cmd/exec.go
> <https://github.com/kubernetes/kubernetes/pull/26693#discussion_r65641725>
> :
>
> >  	containerName := p.ContainerName
> >  	if len(containerName) == 0 {
> >  		glog.V(4).Infof("defaulting container name to %s", pod.Spec.Containers[0].Name)
> >  		containerName = pod.Spec.Containers[0].Name
> >  	}
> >
> > +	isInitContainer := false
> > +	for _, initContainer := range pod.Spec.InitContainers {
> > +		if initContainer.Name == containerName {
> > +			isInitContainer = true
> > +			break
> > +		}
> > +	}
> > +
> > +	if !isInitContainer && pod.Status.Phase != api.PodRunning {
>
> Yeah, you're operating on out-of-date information here. I recommend making
> the attempt and fixing kubelet to send back a great error message if it's
> not possible.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26693/files/bd20e49cd5f40f3c672eab5235a389c97db56fdb#r65641725>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p6lxbHzZWstOSD5cRFoV7e005WrVks5qH3XtgaJpZM4IsWvp>
> .
>